File: windows/GlassTranslate.Windows/Services/SettingsStore.cs
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text.Json;
using GlassTranslate.Windows.Models;
using Microsoft.Win32;

namespace GlassTranslate.Windows.Services;

public sealed class SettingsStore
{
    private const string StartupName = "Wonder";
    private readonly string _folder = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "Wonder");

    private string SettingsPath => Path.Combine(_folder, "settings.json");
    public string FreeServiceEmail { get; set; } = string.Empty;
    public TargetLanguage TargetLanguage { get; set; } = TargetLanguage.Automatic;
    public bool IsPaused { get; set; }
    public bool LaunchAtLogin { get; set; }
    public SelectionMethod SelectionMethod { get; set; } = SelectionMethod.Automatic;
    public double SelectionDelay { get; set; } = 0.18;
    public double BubbleScale { get; set; } = 1.0;
    public BubblePosition BubblePosition { get; set; } = BubblePosition.Below;
    public GlassTheme GlassTheme { get; set; } = GlassTheme.Aurora;
    public double TranslationFontSize { get; set; } = 16;
    public bool ShowSourceText { get; set; } = true;

    public void Load()
    {
        Directory.CreateDirectory(_folder);
        if (File.Exists(SettingsPath))
        {
            try
            {
                var data = JsonSerializer.Deserialize<PersistedSettings>(File.ReadAllText(SettingsPath));
                if (data is not null)
                {
                    FreeServiceEmail = data.FreeServiceEmail ?? string.Empty;
                    TargetLanguage = data.TargetLanguage;
                    IsPaused = data.IsPaused;
                    SelectionMethod = data.SelectionMethod;
                    SelectionDelay = data.SelectionDelay;
                    BubbleScale = data.BubbleScale;
                    BubblePosition = data.BubblePosition;
                    GlassTheme = data.GlassTheme;
                    TranslationFontSize = data.TranslationFontSize;
                    ShowSourceText = data.ShowSourceText;
                }
            }
            catch { /* Keep safe defaults when local settings are damaged. */ }
        }
        LaunchAtLogin = ReadLaunchAtLogin();
    }

    public void Save()
    {
        Directory.CreateDirectory(_folder);
        var json = JsonSerializer.Serialize(
            new PersistedSettings(FreeServiceEmail.Trim(), TargetLanguage, IsPaused,
                SelectionMethod, SelectionDelay, BubbleScale, BubblePosition, GlassTheme, TranslationFontSize, ShowSourceText),
            new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(SettingsPath, json);
        WriteLaunchAtLogin(LaunchAtLogin);
    }

    private static bool ReadLaunchAtLogin()
    {
        using var key = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows\CurrentVersion\Run");
        return key?.GetValue(StartupName) is string;
    }

    private static void WriteLaunchAtLogin(bool enabled)
    {
        using var key = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows\CurrentVersion\Run", writable: true);
        if (key is null) return;
        if (!enabled)
        {
            key.DeleteValue(StartupName, throwOnMissingValue: false);
            return;
        }

        var processPath = Environment.ProcessPath ?? Process.GetCurrentProcess().MainModule?.FileName ?? string.Empty;
        var entryName = Assembly.GetEntryAssembly()?.GetName().Name ?? "Wonder";
        var entryPath = Path.Combine(AppContext.BaseDirectory, $"{entryName}.dll");
        var command = Path.GetFileNameWithoutExtension(processPath).Equals("dotnet", StringComparison.OrdinalIgnoreCase)
            ? $"\"{processPath}\" \"{entryPath}\""
            : $"\"{processPath}\"";
        key.SetValue(StartupName, command);
    }

    private sealed record PersistedSettings(
        string? FreeServiceEmail = "",
        TargetLanguage TargetLanguage = TargetLanguage.Automatic,
        bool IsPaused = false,
        SelectionMethod SelectionMethod = SelectionMethod.Automatic,
        double SelectionDelay = 0.18,
        double BubbleScale = 1.0,
        BubblePosition BubblePosition = BubblePosition.Below,
        GlassTheme GlassTheme = GlassTheme.Aurora,
        double TranslationFontSize = 16,
        bool ShowSourceText = true);
}

File: macos/Sources/GlassTranslate/SettingsStore.swift
import Foundation
import ApplicationServices
import ServiceManagement

@MainActor
final class SettingsStore: ObservableObject {
    static let shared = SettingsStore()

    @Published var targetLanguage: TargetLanguage
    @Published var freeServiceEmail: String
    @Published var launchAtLogin: Bool
    @Published var isPaused: Bool
    @Published var selectionMethod: SelectionMethod
    @Published var selectionDelay: Double
    @Published var bubbleScale: Double
    @Published var bubblePosition: BubblePosition
    @Published var glassTheme: GlassTheme
    @Published var translationFontSize: Double
    @Published var showSourceText: Bool
    @Published var accessibilityGranted: Bool

    private let defaults = UserDefaults.standard
    private init() {
        targetLanguage = TargetLanguage(rawValue: defaults.string(forKey: "targetLanguage") ?? "") ?? .automatic
        freeServiceEmail = defaults.string(forKey: "freeServiceEmail") ?? ""
        isPaused = defaults.bool(forKey: "isPaused")
        selectionMethod = SelectionMethod(rawValue: defaults.string(forKey: "selectionMethod") ?? "") ?? .automatic
        selectionDelay = defaults.object(forKey: "selectionDelay") as? Double ?? 0.18
        bubbleScale = defaults.object(forKey: "bubbleScale") as? Double ?? 1.0
        bubblePosition = BubblePosition(rawValue: defaults.string(forKey: "bubblePosition") ?? "") ?? .below
        glassTheme = GlassTheme(rawValue: defaults.string(forKey: "glassTheme") ?? "") ?? .aurora
        translationFontSize = defaults.object(forKey: "translationFontSize") as? Double ?? 16
        showSourceText = defaults.object(forKey: "showSourceText") as? Bool ?? true
        accessibilityGranted = AXIsProcessTrusted()
        if #available(macOS 13.0, *) {
            launchAtLogin = SMAppService.mainApp.status == .enabled
        } else {
            launchAtLogin = false
        }
    }

    func refreshAccessibilityPermission() {
        accessibilityGranted = AXIsProcessTrusted()
    }

    func save() throws {
        defaults.set(targetLanguage.rawValue, forKey: "targetLanguage")
        defaults.set(freeServiceEmail.trimmingCharacters(in: .whitespacesAndNewlines), forKey: "freeServiceEmail")
        defaults.set(isPaused, forKey: "isPaused")
        defaults.set(selectionMethod.rawValue, forKey: "selectionMethod")
        defaults.set(selectionDelay, forKey: "selectionDelay")
        defaults.set(bubbleScale, forKey: "bubbleScale")
        defaults.set(bubblePosition.rawValue, forKey: "bubblePosition")
        defaults.set(glassTheme.rawValue, forKey: "glassTheme")
        defaults.set(translationFontSize, forKey: "translationFontSize")
        defaults.set(showSourceText, forKey: "showSourceText")
        try updateLaunchAtLogin()
    }

    private func updateLaunchAtLogin() throws {
        if #available(macOS 13.0, *) {
            if launchAtLogin, SMAppService.mainApp.status != .enabled {
                try SMAppService.mainApp.register()
            } else if !launchAtLogin, SMAppService.mainApp.status == .enabled {
                try SMAppService.mainApp.unregister()
            }
        }
    }
}
